File: example-microservice-java/src/main/java/com/example/java/customer/api/CustomerApiMapper.java
package com.example.java.customer.api;

import com.example.java.customer.api.dto.CustomerDto;
import com.example.java.customer.service.CustomerModel;
import org.mapstruct.AfterMapping;
import org.mapstruct.Mapper;
import org.mapstruct.MappingTarget;
import org.mapstruct.factory.Mappers;

import java.util.Optional;

@Mapper
public interface CustomerApiMapper {

    CustomerApiMapper INSTANCE = Mappers.getMapper(CustomerApiMapper.class);

    CustomerDto customerToCustomerDto(CustomerModel source);

    @AfterMapping
    default Optional<CustomerDto> map(Optional<CustomerModel> source, @MappingTarget Optional<CustomerDto> target) {
        return source.map(this::customerToCustomerDto);
    }

}

File: example-microservice-kotlin/src/main/kotlin/com/gooddata/kotlin/customer/api/CustomerApiMapper.kt
package com.example.kotlin.customer.api

import com.example.kotlin.customer.api.dto.CustomerDto
import com.example.kotlin.customer.service.CustomerModel
import org.mapstruct.AfterMapping
import org.mapstruct.Mapper
import org.mapstruct.MappingTarget
import java.util.*

@Mapper
abstract class CustomerApiMapper {

    abstract fun customerToCustomerDto(source: CustomerModel): CustomerDto

    @AfterMapping
    fun map(source: Optional<CustomerModel>, @MappingTarget target: Optional<CustomerDto>): Optional<CustomerDto> {
        return source.map { customerToCustomerDto(it) }
    }

}